lib/Mojolicious/Plugin/ClientIP/Pluggable.pm

```diff
@@ -0,0 +1,215 @@
+(rfc-7239), or use C<remote_address> environment.
+
+The pluging is inspired by L<Mojolicious::Plugin::ClientIP>.
+
+=head1 METHODS
+
+=head2 client_ip
+
+Find a client IP address from the specified headers, with optional fallbacks. The address is
+validated that it is publicly available (aka routable) IP address. Empty string is returned
+if no valid address can be found.
+
+=head1 OPTIONS
+
+=head2 analyzed_headers
+
+Define order and names of cloud provider injected headers with client IP address.
+For C<cloudflare> we found the following headers are suitable:
+
+    plugin 'ClientIP::Pluggable',
+        analyzed_headers => [qw/cf-pseudo-ipv4 cf-connecting-ip true-client-ip/].
+
+This option is mandatory.
+
+More details at L<https://support.cloudflare.com/hc/en-us/articles/202494830-Pseudo-IPv4-Supporting-IPv6-addresses-in-legacy-IPv4-applications>,
+L<https://support.cloudflare.com/hc/en-us/articles/200170986-How-does-CloudFlare-handle-HTTP-Request-headers>,
+L<https://support.cloudflare.com/hc/en-us/articles/206776727-What-is-True-Client-IP>
+
+=head2 restrict_family
+
+    plugin 'ClientIP::Pluggable', restrict_family => 'ipv4';
+    plugin 'ClientIP::Pluggable', restrict_family => 'ipv6';
+
+If defined only IPv4 or IPv6 addresses are considered valid among the possible addresses.
+
+By default this option is not defined, allowing IPv4 and IPv6 addresses.
+
+=head2 fallbacks
+
+    plugin 'ClientIP::Pluggable',
+        fallbacks => [qw/rfc-7239 x-forwarded-for remote_address/]);
+
+Try to get valid client IP-address from fallback sources, if we fail to do that from
+cloud-provider headers.
+
+C<rfc-7239> uses C<forwarded> header, C<x-forwarded-for> use <x-forwarded-for> header
+(appeared before rfc-7239 and still widely used) or use remote_address environment
+(C<$c->tx->remote_address>).
+
+Default value is C<[remote_address]>.
+
+=head1 ENVIRONMENT
+
+=head2 CLIENTIP_PLUGGABLE_ALLOW_LOOPBACK
+
+Allows non-routable loopback address (C<127.0.0.1>) to pass validation. Use it for
+test purposes.
+
+Default value is C<0>, i.e. loopback addresses do not pass IP-address validation.
+
+
+=head1 COPYRIGHT AND LICENSE
+
+Copyright (C) 2017 binary.com
+
+=cut
+
+
+use strict;
+use warnings;
+
+use Data::Validate::IP;
+
+use Mojo::Base 'Mojolicious::Plugin';
+
+our $VERSION = '0.01';
+
+# for tests only
+use constant ALLOW_LOOPBACK => $ENV{CLIENTIP_PLUGGABLE_ALLOW_LOOPBACK} || 0;
+
+sub _check_ipv4 {
+    my ($ip) = @_;
+    return Data::Validate::IP::is_public_ipv4($ip)
+        || (ALLOW_LOOPBACK && Data::Validate::IP::is_loopback_ipv4($ip));
+}
+
+sub _check_ipv6 {
+    my ($ip) = @_;
+    return Data::Validate::IP::is_public_ipv6($ip)
+        || (ALLOW_LOOPBACK && Data::Validate::IP::is_loopback_ipv6($ip));
+}
+
+sub _classify_ip {
+    my ($ip) = @_;
+    return Data::Validate::IP::is_ipv4($ip)
+        ? 'ipv4'
+        : Data::Validate::IP::is_ipv6($ip)
+        ? 'ipv6'
+        : undef;
+}
+
+sub _candatetes_iterator {
```

Comment:
may be rename it to better `_candidates_iterator`, i don't know what `candatetes` means